src/bin/mnemosyne/fat/layout/layout-derive/src/lib.rs

```diff
@@ -0,0 +1,239 @@
+                let all_name = Ident::new(&format!("all_{name}"), name.span());
+                quote! {
+                    pub fn #name (&mut self) -> Result<<#ty as ::layout::ApplyLayout<'_, R>>::Frame, R::Error> {
+                        let offset_in_parent = #offset_in_parent;
+                        <#ty as ::layout::ApplyLayout<'_, R>>::apply_layout(self.stream, self.offset + offset_in_parent)
+                    }
+
+                    pub fn #all_name (&mut self) -> Result<#ty, R::Error> {
+                        let offset_in_parent = #offset_in_parent;
+                        self.stream.seek(::layout::io::SeekFrom::Start(self.offset + offset_in_parent))?;
+                        <#ty as ::layout::Decode>::decode(self.stream)
+                    }
+                }
+            }
+            FieldLayout::Primitive => quote! {
+                pub fn #name (&mut self) -> Result<#ty, R::Error> {
+                    let offset_in_parent = #offset_in_parent;
+                    self.stream.seek(::layout::io::SeekFrom::Start(self.offset + offset_in_parent))?;
+                    <#ty as ::layout::Decode>::decode(self.stream)
+                }
+            },
+        }
+    }
+
+    fn gen_setter(&self, offset_in_parent: &TokenStream2, name: &Ident, ty: &Type) -> TokenStream2 {
+        let setter_name = Ident::new(&format!("set_{name}"), name.span());
+        quote! {
+            pub fn #setter_name (&mut self, data: &#ty) -> Result<(), W::Error> {
+                let offset_in_parent = #offset_in_parent;
+                self.stream.seek(::layout::io::SeekFrom::Start(self.offset + offset_in_parent))?;
+                <#ty as ::layout::Encode>::encode(data, self.stream)
+            }
+        }
+    }
+}
+
+enum Size {
+    Dynamic,
+    Fixed,
+}
+
+impl Size {
+    fn gen_field_size(&self, name: &Ident, ty: &Type) -> [TokenStream2; 3] {
+        match self {
+            Size::Dynamic => [
+                quote!(::layout::FramedDynamic::framed_size(&mut self.#name()?)?),
+                quote!(::layout::SourcedDynamic::sourced_size(&#name)),
+                quote!(::layout::SourcedDynamic::sourced_size(&self.#name)),
+            ],
+            Size::Fixed => {
+                let s = quote!(<#ty as ::layout::Fixed>::size());
+                [s.clone(), s.clone(), s]
+            }
+        }
+    }
+
+    fn gen_impl_wrapper(
+        &self,
+        framed: TokenStream2,
+        sourced: TokenStream2,
+        name: &Ident,
+        frame_name: &Ident,
+    ) -> TokenStream2 {
+        match self {
+            Size::Dynamic => quote! {
+                impl<'a, S> ::layout::FramedDynamic<S> for #frame_name<'a, S>
+                where
+                    S: ::layout::Read + ::layout::Seek + ::layout::IO
+                {
+                    fn framed_size(&mut self) -> Result<u64, S::Error> {
+                        Ok(#framed)
+                    }
+                    
```

Comment:
There is a lot of extra white space throuhout.  Can we tighten this up?  Perhaps @dbittman has thoughts on how the code looks as well?

Reply:
One thing we need to do is setup a CI step that ensures that rustfmt is happy. @CPTforever is this code formatted with rustfmt?

Reply:
I'm pretty sure I did, I ran cargo fmt in the crate's root but I'll double check.

Reply:
Ok today I found out it's not recursive, I'll commit the fix soon,